server/subscribe.go

```diff
@@ -933,6 +1091,97 @@ func (p *Plugin) httpChannelGetSubscriptions(w http.ResponseWriter, r *http.Requ
 	return respondJSON(w, subscriptions)
 }
 
+func (p *Plugin) httpGetSubscriptionTemplates(w http.ResponseWriter, r *http.Request, mattermostUserID string) (int, error) {
+	instanceID := types.ID(r.FormValue("instance_id"))
+
+	if len(instanceID) < 2 {
+		return respondErr(w, http.StatusBadRequest,
+			errors.New("bad or missing instance id"))
+	}
+
+	subscriptionTemplates, err := p.getSubscriptionTemplatesForInstance(instanceID)
+	if err != nil {
+		return respondErr(w, http.StatusInternalServerError,
+			errors.Wrap(err, "unable to get channel subscription templates"))
+	}
+
+	client, _, _, err := p.getClient(instanceID, types.ID(mattermostUserID))
+	if err != nil {
+		return respondErr(w, http.StatusInternalServerError, err)
+	}
+
+	plist, err := client.ListProjects("", -1)
+	if err != nil {
+		return respondErr(w, http.StatusInternalServerError, err)
+	}
+
+	var subTemplates []SubscriptionTemplate
+	for _, project := range plist {
+		listSubscriptionTemplate := subscriptionTemplates.Templates.ByProjectID[project.Key]
+		for _, subTemplate := range listSubscriptionTemplate {
+			subTemplates = append(subTemplates, subTemplate)
+		}
+	}
+	return respondJSON(w, subTemplates)
+}
+
+func (p *Plugin) httpCreateSubscriptionTemplate(w http.ResponseWriter, r *http.Request, mattermostUserID string) (int, error) {
+	subscriptionTemplate := SubscriptionTemplate{}
+	err := json.NewDecoder(r.Body).Decode(&subscriptionTemplate)
+	if err != nil {
+		return respondErr(w, http.StatusBadRequest,
+			errors.WithMessage(err, "failed to decode incoming request"))
+	}
+
+	client, _, _, err := p.getClient(subscriptionTemplate.InstanceID, types.ID(mattermostUserID))
+	if err != nil {
+		return respondErr(w, http.StatusInternalServerError, err)
+	}
+
+	if err = p.addSubscriptionTemplate(subscriptionTemplate.InstanceID, &subscriptionTemplate, client); err != nil {
+		return respondErr(w, http.StatusInternalServerError, err)
+	}
+
+	code, err := respondJSON(w, &subscriptionTemplate)
+	if err != nil {
+		return code, err
+	}
+
+	return http.StatusOK, nil
+}
+
+func (p *Plugin) httpDeleteSubscriptionTemplate(w http.ResponseWriter, r *http.Request, mattermostUserID string) (int, error) {
+	subscriptionTemplateID := strings.TrimPrefix(r.URL.Path, routeAPISubscriptionTemplates+"/")
+	if len(subscriptionTemplateID) != 26 {
+		return respondErr(w, http.StatusBadRequest,
+			errors.New("bad subscription id"))
```

Comment:
are these lines that much large that we need to write them in separate lines? if not, fix everywhere